### ISSUE TITLE
New Check: Resource and data source parameters should include a Description field

### ISSUE THREAD

**Issue by gdavison** [bflad/tfproviderlint#35]
The `Description` field on a parameter can be used to generate documentation and can be used e.g. for tooling like a language server.

**Comment by bflad** [comment]
This would be an excellent addition! 😄 The code for this check certainly deserves to live here in this codebase.

My only concern would be whether this should be enforced by default. Our current checks are fairly universally accepted practices (and therefore all enabled by default in the `tfproviderlint` release binary), however introducing a new `Description` field burden on across all attributes on provider maintainers when the automatic documentation generation tooling is not there yet feels a little heavy handed at the current moment. At some point in the future, this likely would be enabled by default though.

We do not have a precedent for how to deal with "optional"/"advanced"/"experimental"/"beta" checks yet (e.g. naming them, disable/enabling them) so there is likely some design or proof of concept work that needs to happen first. 👍  I think `golang.org/x/tools/go/analysis/multichecker` may require creating a second release binary as the only way to setup different default checks.

**Comment by gdavison** [comment]
Yes, there would definitely be a lot of design work :) And it should be an opt-in check.

One additional thing could be adding a feature like [RuboCop](https://www.rubocop.org/en/stable/)'s [`--auto-gen-config` flag](https://rubocop.readthedocs.io/en/latest/configuration/#automatically-generated-configuration) which can create a TODO file from existing violations

**Comment by bflad** [comment]
This new opt-in check has been implemented via #67 -- a small sample of its results against the Terraform AWS Provider includes:

```
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_workspaces_ip_group.go:23:12: XS001: schema should configure Description
22			Schema: map[string]*schema.Schema{
23				"name": {
24					Type:     schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_workspaces_ip_group.go:28:19: XS001: schema should configure Description
27				},
28				"description": {
29					Type:     schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_workspaces_ip_group.go:33:13: XS001: schema should configure Description
32				},
33				"rules": {
34					Type:     schema.TypeSet,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_workspaces_ip_group.go:38:17: XS001: schema should configure Description
37						Schema: map[string]*schema.Schema{
38							"source": {
39								Type:     schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_workspaces_ip_group.go:42:22: XS001: schema should configure Description
41							},
42							"description": {
43								Type:     schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:25:17: XS001: schema should configure Description
24			Schema: map[string]*schema.Schema{
25				"rule_name": {
26					Type:         schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:31:20: XS001: schema should configure Description
30				},
31				"resource_arn": {
32					Type:     schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:35:16: XS001: schema should configure Description
34				},
35				"priority": {
36					Type:         schema.TypeInt,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:40:18: XS001: schema should configure Description
39				},
40				"fixed_rate": {
41					Type:     schema.TypeFloat,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:44:22: XS001: schema should configure Description
43				},
44				"reservoir_size": {
45					Type:         schema.TypeInt,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:49:20: XS001: schema should configure Description
48				},
49				"service_name": {
50					Type:         schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:54:20: XS001: schema should configure Description
53				},
54				"service_type": {
55					Type:         schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:59:12: XS001: schema should configure Description
58				},
59				"host": {
60					Type:         schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:64:19: XS001: schema should configure Description
63				},
64				"http_method": {
65					Type:         schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:69:16: XS001: schema should configure Description
68				},
69				"url_path": {
70					Type:         schema.TypeString,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:74:15: XS001: schema should configure Description
73				},
74				"version": {
75					Type:         schema.TypeInt,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:80:18: XS001: schema should configure Description
79				},
80				"attributes": {
81					Type:     schema.TypeMap,
/Users/bflad/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_xray_sampling_rule.go:88:11: XS001: schema should configure Description
87				},
88				"arn": {
89					Type:     schema.TypeString,
```

The process of documenting how to enable these extra checks, which live in a separate package path (`xpasses`), are prefixed with `X`, and are not included with the default `tfproviderlint` binary, will be covered for downstream linters via #68. Any enhancements to include these checks in the default binary in a disabled state should be a separate issue, but I believe it will require upstream `go/analysis` improvements and I'd really like to avoid straying too far outside the default functionality provided by the framework in this project.

**Comment by bflad** [comment]
Opting for now to also include a quick `tfproviderlintx` command in addition to the instructions for how to implement a custom lint tool: https://github.com/bflad/tfproviderlint/pull/69
